ceph/rados/serviceability_workflows.py

```diff
@@ -315,6 +315,87 @@ def remove_osds_from_host(self, host_obj):
                     service_type="osd", service_name=service
                 )
 
+    def add_osds_to_managed_service(self, osds: list = None, spec: str = None) -> bool:
+        """
+        Method to identify non-managed OSDs and then add them back to managed state under existing OSD specs
+        Args:
+            osds: list of OSDs to be set to managed state
+            spec: name of the spec that needs to be used for setting OSD to manage.
+        Returns:
+            Pass -> True
+            fail -> False
+
+        Note : if osds list is provided, Only those osds will be added to managed and check if the cluster,
+        is free of unmanaged OSDs.
+
+        Note : If no OSDs are provided, we add all OSDs to managed host by host,
+        until we're left with no unmanaged osds on cluster.
+
+        Note: If Spec is not provided, 1st available - managed OSD spec name would be used
+        """
+
+        def get_osd_services():
+            cmd_export = "ceph orch ls osd --export"
+            return self.rados_obj.run_ceph_command(cmd=cmd_export, client_exec=True)
+
+        def check_unmanaged_osds_exist():
+            out = get_osd_services()
+            for _service in out:
+                if not (
+                    _service.get("placement", False)
+                    or _service.get("service_id", False)
+                ):
+                    log.info("Service %s not managed", _service)
+                    return True
+            log.info("All OSD daemons are managed by named service")
+            return False
+
+        if not check_unmanaged_osds_exist():
+            log.info("No unmanaged OSDs exist. Exiting")
+            return True
+
+        cluster_managed_osd_specs = [
+            name["service_name"]
+            for name in get_osd_services()
+            if ("service_name" and "service_id") in name
+        ]
+        if spec and spec not in cluster_managed_osd_specs:
+            log.error("Passed spec does not exist on the cluster")
+            return False
+
+        def set_osds_managed(osd_list):
+            set_managed_raw = "ceph orch osd set-spec-affinity"
+            osd_args = " ".join(str(osd) for osd in osd_list)
+            final_cmd = f"{set_managed_raw} {spec if spec else cluster_managed_osd_specs[0]} {osd_args}"
+            self.rados_obj.run_ceph_command(cmd=final_cmd)
+
+        if osds:
+            set_osds_managed(osds)
+            time.sleep(10)
+            if check_unmanaged_osds_exist():
+                log.error(
+                    "Unmanaged OSDs exist on cluster even after adding the provided osds"
+                )
+                return False
```

Comment:
This would fail if cluster is having 5 unmanaged OSDs and user is trying to mode only 3 OSDs to managed state.

Reply:
Yes, when user passes the OSD IDs, we're expecting it to clear up the cluster of non-managable OSDs. That's the idea.

If this not needed, we can update the workflow.

Reply:
Updated the behaviour. Now failing only if the OSD passed does not have a spec affinity post movement